mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java

```diff
@@ -880,22 +867,21 @@ public void onData(ByteBuffer data, boolean endStream, EnvoyStreamIntel streamIn
         return;
       }
 
-      Runnable task = new Runnable() {
-        @Override
-        public void run() {
-          checkCallingThread();
-          try {
-            ByteBuffer userBuffer = mUserCurrentReadBuffer;
-            mUserCurrentReadBuffer = null; // Avoid the reference to a potentially large buffer.
-            int dataRead = data.remaining();
-            userBuffer.put(data); // NPE ==> BUG, BufferOverflowException ==> User not behaving.
-            if (dataRead > 0 || !endStream) {
-              mWaitingOnRead.set(true);
-              mCallback.onReadCompleted(CronvoyUrlRequest.this, mUrlResponseInfo, userBuffer);
-            }
-          } catch (Throwable t) {
-            onCallbackException(t);
+      ByteBuffer userBuffer = mUserCurrentReadBuffer;
+      mUserCurrentReadBuffer = null; // Avoid the reference to a potentially large buffer.
+      int dataRead = data.remaining();
+      // Copy the `data` outside the thread before passing into the thread because the `data`
+      // will be destroyed upon completing this callback.
+      userBuffer.put(data); // NPE ==> BUG, BufferOverflowException ==> User not behaving.
```

Comment:
why did we pull these few lines out of the Runnable?

Reply:
The `data` may get destroyed if we do the copy inside the thread since the callback may complete first.